File: 2025/NeetCode/Arrays & Hashing/242. Valid Anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {

    if (s.length != t.length) return false;

    let map = {};

    for (let i = 0; i < s.length; i++) {
        let letter = s[i];

        if (!map[letter]) {
            map[letter] = 1;
        } else {
            map[letter]++;
        }
    }

    for (let i = 0; i < t.length; i++) {
        let letter = t[i];

        if(map[letter] === undefined) {
            return false;
        } 

        map[letter]--;

        if(map[letter] < 0) {
            return false;
        }

    }
    return true;

};

// Time Complexity = O(n + m)
// Space Complexity = O(n)

File: 2026/Blind 75 [Python]/Arrays & Hashing/242.Valid Anagram.py
from collections import defaultdict

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:

        s_map = defaultdict(int)
        t_map = defaultdict(int)

        for char in s:
            s_map[char] += 1        

        for char in t: 
            t_map[char] += 1
        
        return (t_map == s_map)

# Time Complexity: O(N)
# Space Complexity: O(1) - only 26 keys are possible